### DISCUSSION TITLE
429 Too many requests

### DISCUSSION THREAD

**Discussion by danielgangl** [wevm/wagmi#1146] [General]
Hey guys, I am having trouble with wagmi making waay too many calls implicitly (about 100 eth_getBalance requests on page load).
Are any of you having the same problem?
I am using wagmi with rainbow kit.

<img width="313" alt="Screenshot 2022-10-23 at 22 24 29" src="https://user-images.githubusercontent.com/46197733/197416475-e21010a9-8db4-4734-8d2e-7d7bf0a7d703.png">

**Comment by jxom** [comment]
You may need to provide your own Alchemy API Key to `alchemyProvider`.

**Reply by danielgangl** [reply]
Yeah I did that...

But even then these are way too many requests. Having even a few users on my we app I will instantly be rate-limited and their experience might be impacted negatively

**Reply by Fabien-Morrow** [reply]
You have to probably check your app architecture, 100 hooks re-render for getting a balance is way to much.
For me it's not a wagmi problem. I recommend you to read [this ](https://www.developerway.com/posts/react-re-renders-guide)

**Comment by soccerarenadev** [comment]
I'm getting that error, but it is not due to too many request.
This is the actual message:
`{"message":"Method Not Allowed","logref":null,"path":null,"_links":{"self":{"href":"/v2/asdfasdfasdf","templated":false,"profile":null,"deprecation":null,"title":null,"hreflang":null,"type":null,"name":null}},"_embedded":{"errors":[{"message":"Method [GET] not allowed for URI [/v2/asdfasdfasdf]. Allowed methods: [POST]","logref":null,"path":null,"_links":{},"_embedded":{}}]}}
`
More info: https://github.com/wagmi-dev/wagmi/issues/1133

**Comment by danielgangl** [comment]
I just noticed that the issue does not happen anymore when I remove rainbow kit - so thats more likely the problem. 
Thank you for your answers!

**Reply by soccerarenadev** [reply]
> I just noticed that the issue does not happen anymore when I remove rainbow kit - so thats more likely the problem. Thank you for your answers!

Still shouldn't happen in Wagmi as it should fallback to the next provider instead of failing.

**Reply by dev-cz** [reply]
> I just noticed that the issue does not happen anymore when I remove rainbow kit - so thats more likely the problem. Thank you for your answers!

I also have this issue with Rainbow. Hmmm also have my own Infura Key and it burned right through it.

**Reply by REPTILEHAUS** [reply]
Is there any more info about this and Rainbowkit

**Comment by rsodre** [comment]
I have the same problems here (429 and Method not allowed), and I'm working on a local ganache server.
I can get a key, but I'm afraid to burn it.
Why would wagmi be so heavy on mainnet when working locally?

<img width="626" alt="Screen Shot 2022-10-25 at 19 22 51" src="https://user-images.githubusercontent.com/229161/197893452-7a8aad13-c1e4-41ed-ac46-18127f120630.png">

**Reply by tmm** [reply]
It shouldn't taking a look into it!

**Reply by rsodre** [reply]
Do wagmi still use the publicprovider, or configured provider, after I connect to MetaMask?
I never fetch anything before connecting to MetaMask.

**Comment by dev-cz** [comment]
Did someore digging into this, and it does seem like a wagmi issue.
 
If I don't use alchemy and provide an Infura Key, the issues is still there and prevents even Infura from being used, and actually spams the Infura API calls.

So far what I have found work is to add an Alchemy key, and Infura Key then set priority and that fixes the issue.

Issues has only been there last few weeks so not sure what changed.

**Reply by rsodre** [reply]
Like this? Seems to be working normally again here.

```
const { provider, webSocketProvider } = configureChains(chains, [
	infuraProvider({
		apiKey: process.env.INFURA_API_KEY,
		priority: 0,
	}),
	alchemyProvider({
		apiKey: process.env.ALCHEMY_API_KEY,
		priority: 1,
	}),
	publicProvider({
		priority: 2,
	}),
]);
```

**Reply by dev-cz** [reply]
> Like this? Seems to be working normally again here.
> 
> ```
> const { provider, webSocketProvider } = configureChains(chains, [
> 	infuraProvider({
> 		apiKey: process.env.INFURA_API_KEY,
> 		priority: 0,
> 	}),
> 	alchemyProvider({
> 		apiKey: process.env.ALCHEMY_API_KEY,
> 		priority: 1,
> 	}),
> 	publicProvider({
> 		priority: 2,
> 	}),
> ]);
> ```

Yep! That's how I have it and how I was able to work around the issue until wagmi team has a fix.

**Reply by jxom** [reply]
You don't need to add a priority as `wagmi` does that internally. An API key is enough!

[There is an open PR](https://github.com/wagmi-dev/wagmi/pull/1176) to migrate away from the public Alchemy RPC URLs in `publicProvider`.

**Reply by dev-cz** [reply]
> You don't need to add a priority as `wagmi` does that internally. An API key is enough!
> 
> [There is an open PR](https://github.com/wagmi-dev/wagmi/pull/1176) to migrate away from the public Alchemy RPC URLs in `publicProvider`.

Correct, but if your goal was to use Infura, then you do need the priority, but in any case you could use either at that point.

**Reply by jxom** [reply]
You don't need priority if it is listed first in the array.

**Reply by johnpaulcas** [reply]
> You don't need priority if it is listed first in the array.

Hi @jxom so are we gonna set the publicProvider to be the first in the list? also when infura reach the limit it seems it doesnt change the provider

**Comment by a-znk** [comment]
I'm experiencing a similar issue with my project as well! I didn't make a single change to it, started back on developing a few days later, and BAM, exhibiting this behavior now.

**Reply by dev-cz** [reply]
See my post above. Get an Alchemy Key, and set it in the wagmi config to overcome the issue. Only way I was able to get around it with or without Rainbow.

**Reply by rsodre** [reply]
To me started a few days ago also.

I would even request to optionally use **NO KEY at all**, not even public, just connected provider.

**Reply by a-znk** [reply]
Thank you! I'm building my project using NextJs, and I've found that I need to add "NEXT_PUBLIC" to my environment variable in this instance. Otherwise, I receive this message "Error: could not detect network (event="noNetwork", code=NETWORK_ERROR, version=providers/5.7.2)". 

As a result, my Alchemy API key is exposed in DevTools. I was thinking perhaps I could find a way to make that API call in an API route, but wasn't sure if/how that might work. Any suggestions on how to add this key more securely?

**Reply by rsodre** [reply]
I use next.js also and it does not expose it when I configure it just like the Getting started tutorial.
But when that happens, a simple workaround  is to create an api endpoint on your server, and call the external api from there

**Reply by Fabien-Morrow** [reply]
I'm stuck with this too :
- If I put NEXT_PUBLIC to my env variables, wagmi can properly initialise the providers but they are exposed client side.
- If I remove NEXT_PUBLIC, wagmi can't initialise this client side for calling the hooks and switch to default provider, with very low rate limit.

How did you manage this @rsodre ? Do you have an _app to paste ?

**Reply by tmm** [reply]
> If I put NEXT_PUBLIC to my env variables, wagmi can properly initialise the providers but they are exposed client side.

They are publishable keys and need to be in the JS bundle to work.

**Reply by rsodre** [reply]
> How did you manage this @rsodre ? Do you have an _app to paste ?

[API routes in nextjs](https://nextjs.org/docs/api-routes/introduction) are always executed on the server.
If I need anything from Infura, for example, I have a hidden route like `/api/infura/getbalance` that will do the actual fetching.

Or use [server side props](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props), also guarantee to run on the server

**Comment by gpxl-dev** [comment]
I think by default retries are very agressive here too. It would be ideal if we could listen for 429s and throttle back retries in those events.

This also relates to this request - please throw on an upvote!
https://github.com/wagmi-dev/wagmi/discussions/950

**Reply by jxom** [reply]
Yeah that would be nice, unfortunately we don't have control over the retry timing. It internally uses ethers' exponential backoff algorithm for retries.

**Reply by gpxl-dev** [reply]
Ah that makes sense, good to know!

**Comment by thekesslereffect** [comment]
Having the same 429 issue as everyone else. Any fix for it yet?

**Comment by newbluestudios** [comment]
Also having this issue with Rainbowkit.  Can't figure out what went wrong.  The docs don't have all the info to fix the issue either 😭

**Comment by johnpaulcas** [comment]
Im having same issue, 429 im using connectKit, I just add new alchecmy and this error occur even on the first few request

**Reply by newbluestudios** [reply]
for me this fixed it (above)

const { provider, webSocketProvider } = configureChains(chains, [
	infuraProvider({
		apiKey: process.env.INFURA_API_KEY,
		priority: 0,
	}),
	alchemyProvider({
		apiKey: process.env.ALCHEMY_API_KEY,
		priority: 1,
	}),
	publicProvider({
		priority: 2,
	}),
]);

**Reply by johnpaulcas** [reply]
> for me this fixed it (above)
> 
> const { provider, webSocketProvider } = configureChains(chains, [ infuraProvider({ apiKey: process.env.INFURA_API_KEY, priority: 0, }), alchemyProvider({ apiKey: process.env.ALCHEMY_API_KEY, priority: 1, }), publicProvider({ priority: 2, }), ]);

thanks! will try this one

**Comment by johnpaulcas** [comment]
anyway to like used, the metamask as default provider if connected? than using alchemy or infura as provider as 429 occur a lot, hope to hear from the team soon, thanks!

**Reply by gpxl-dev** [reply]
Upvote #950

**Comment by arthur-zhuk** [comment]
Someone mentioned that the issue goes away if you remove `process.env` from the evaluation and just hard code the api key as a string. That works, but, why is this bug suddenly alive and hasn't been addressed?

**Reply by johnpaulcas** [reply]
same! maybe if they know some solution, they can provide it to docx and wondering if than just using alchemy or infura as provider, we can just get the default connected wallet like metamask for example as default provider.

**Comment by rafinskipg** [comment]
I found the problem people.

Alchemy provider accepts one "api_key" , that is used for several networks. But that doesn't work!

Each API_KEY in alchemy is only for one network. 

So this setup will work for the default network you created your apikey for , but not for others. 

I think it needs a rewrite on wagmi alchemy provider

**Reply by tmm** [reply]
Unless something changed on their end, Alchemy API keys should work across networks (it only looks like in the dashboard that they don't).

**Reply by rafinskipg** [reply]
yes you are right i was confused, trying to find a solution xDD

**Comment by ackvf** [comment]
This sucks as it seems to ignore the setting for `stallTimeout`. This is my config
```ts
const { chains, provider, webSocketProvider } = configureChains(
  getSupportedChains(),
  [
    publicProvider({
      priority: 0,
      stallTimeout: 1000,
    }),
    alchemyProvider({
      priority: 1,
      apiKey: ALCHEMY_API_KEY,
    }),
  ],
  {
    stallTimeout: 1000,
  }
)
```
This says that after a second it should switch from `publicProvider` to `alchemyProvider`.

I do indeed see a request with my own alchemy key after a second, which retrieves the expected result, but the public provider still keeps firing even after a minute. WTF??

<img width="338" alt="image" src="https://user-images.githubusercontent.com/2313018/201232257-93aa671c-c81e-43d5-ac60-b98827a16ff8.png">


Changing the priority from public:0 alchemy:1 to public:1 alchemy:0 solves the issue and there are no longer this many requests, but just one.

**Comment by ackvf** [comment]
Just a heads up guys, the 429 Issue was solved in latest versions of wagmi where they changed the default RPC url for the troubling chains.
Just upgrade to latest wagmi and you should be K.


/edit The issue was that the Alchemy url with an actual API key was hardcoded in previous versions of wagmi, ignoring whatever you set.

**Reply by johnpaulcas** [reply]
@ackvf just clarification, does that mean you need to upgrade to >0.8.0, and how did it get fixed if the connected wallet rpc is not being used? just wondering 💭

also for verification @tmm 

thanks 🙏

**Reply by tmm** [reply]
Fixed in [`wagmi@0.7.8`](https://github.com/wagmi-dev/wagmi/releases/tag/wagmi%400.7.8)

**Reply by johnpaulcas** [reply]
hi @tmm, I upgraded from 0.7.15 to 0.7.8 and it got worst, I encounter the isDataEqual issue again + 429 😅, this already fixed on version 0.7.15

<img width="531" alt="Screen Shot 2022-11-17 at 8 27 17 AM" src="https://user-images.githubusercontent.com/44362950/202324262-4c1f4be4-ebfe-4fcd-8541-10729eba452d.png">

**Reply by ackvf** [reply]
I upgraded straight to 0.8.1 few days ago ¯\\_(ツ)_/¯ and the issue was gone. Maybe it was a coincidence, but worth a shot upgrading.

**Reply by ackvf** [reply]
@johnpaulcas the api key was hardcoded in the chain configuration, look here https://github.com/wagmi-dev/wagmi/commit/389765f7d9af063ab0df07389a2bbfbc10a41060 

This was the one failing for everyone `https://eth-goerli.alchemyapi.io/v2/_gg7wSSi0KMBsdKnGVfHDueq6xMB9EkC`. You can even search it with google `_gg7wSSi0KMBsdKnGVfHDueq6xMB9EkC` and find many related issues.


But what I am seeing in your request url is that you are using different key, so maybe you are using your api key and have reached the quota on it too?

**Reply by johnpaulcas** [reply]
@ackvf actually thats api is just newly added, Im using a loop and inside of loop are `watch` option enabled, maybe because consecutive call in alchemy is not allowed, but when I switch it to infura, it works just fine

**Reply by johnpaulcas** [reply]
weird, I switch the ordering and now my alchemy works 🤔

<img width="515" alt="Screen Shot 2022-11-17 at 11 43 44 AM" src="https://user-images.githubusercontent.com/44362950/202349773-fce4bf1b-9ec9-468a-9c98-c0f71a91ad68.png">

**Reply by rsodre** [reply]
I also still keep seeing the error all the time, with the public key.

I import ethers on some components to use ethers.utils, should that be the cause?

`const { ethers } = require('ethers');`

What is the proper way to import ethers with wagmi?

**Reply by rsodre** [reply]
Nevermind, my keys were not visible.

**Comment by Hussainzz** [comment]
it's weird latest version of WAGMI seems to throw a rate limit 429 error with alchemyProvider.
but it works fine with `"wagmi": "^0.8.8"`

```
const { chains, provider, webSocketProvider } = configureChains(
  [polygon, polygonMumbai],
  [
    alchemyProvider({ apiKey: process.env.REACT_APP_ALCHEMY_API_KEY }),
    publicProvider(),
  ]
);
```

**Comment by devlancer412** [comment]
I didn't use alchemy api. but it sends too many request and all response are 429 error.
Please help me.

```

const { chains, provider } = configureChains(
  [flare, ...(process.env.NODE_ENV === "development" ? [coston2] : [])],
  [
    jsonRpcProvider({
      rpc: (chain) => ({
        http: RPC_URLS[chain.id],
      }),
    }),
  ],
  { pollingInterval: 10_000, targetQuorum: 3 },
);

export const client = createClient({
  autoConnect: true,
  connectors: [
    new MetaMaskConnector({ chains }),
    new CoinbaseWalletConnector({
      chains,
      options: {
        appName: "wagmi",
      },
    }),
    new WalletConnectConnector({
      chains,
      options: {
        qrcode: true,
      },
    }),
  ],
  provider,
});

```

**Reply by Stenor5** [reply]
> I didn't use alchemy api. but it sends too many request and all response are 429 error. Please help me.
> 
> ```
> 
> const { chains, provider } = configureChains(
>   [flare, ...(process.env.NODE_ENV === "development" ? [coston2] : [])],
>   [
>     jsonRpcProvider({
>       rpc: (chain) => ({
>         http: RPC_URLS[chain.id],
>       }),
>     }),
>   ],
>   { pollingInterval: 10_000, targetQuorum: 3 },
> );
> 
> export const client = createClient({
>   autoConnect: true,
>   connectors: [
>     new MetaMaskConnector({ chains }),
>     new CoinbaseWalletConnector({
>       chains,
>       options: {
>         appName: "wagmi",
>       },
>     }),
>     new WalletConnectConnector({
>       chains,
>       options: {
>         qrcode: true,
>       },
>     }),
>   ],
>   provider,
> });
> ```

I have same issue while using wagmi.

**Comment by firnprotocol** [comment]
related question: how do we "respect" the user's choice of custom RPC?

i.e., let's say that the user manually adds a **custom chain** into MetaMask, with Optimism's chainID, but with a **custom RPC URL** (e.g., pointing to _their own_ alchemy).

i am using the standard `publicProvider`. it appears that in this setting wagmi still sends connections through the public Optimism gateway, `mainnet.optimism.io`. i.e., it's ignoring the RPC url set in the wallet extension, and instead using something hardcoded in wagmi.

how do we respect / defer to the (custom) RPC url set in MetaMask? thanks in advance.

**Reply by tmm** [reply]
See https://github.com/wagmi-dev/wagmi/discussions/950#discussioncomment-4200440

**Reply by firnprotocol** [reply]
thanks for the heads-up. it looks like it _is possible_, à la [this comment](https://github.com/wagmi-dev/wagmi/discussions/950#discussioncomment-4207884)?

my main motivation here is privacy. as a privacy-oriented application, i want to defer to the user's choice of RPC, if they have reason to use something custom. (there are plenty of reasons why one might want to use a custom RPC.)

**Reply by jxom** [reply]
Probably. But it is pretty much impossible to guarantee determinism if you rely on the user to set their own RPC, which may lead to many bugs in your dapp. The RPC may lack compatibility, become compromised, or have poor rate quotas. These problems don’t exist if the dapp can sensibly choose which RPC(s) they want to route through.

**Reply by firnprotocol** [reply]
i certainly agree with you that there are cons; on the other hand, there are also pros (see my comment). so it's a matter of tradeoffs. in any case, for my future reference, it'd be helpful to know if there was a "minimal" / easy way to achieve this functionality, just so that i can test it? something like [this approach](https://github.com/wagmi-dev/wagmi/discussions/950#discussioncomment-4207884)? keep in mind my code only uses `configureChains` and `createConfig`, but not `createClient`; this appears to be the "new" way of doing things. thanks in advance for your patience.

**Comment by REPTILEHAUS** [comment]
Im not sure if this is related but im getting this when metamask popup opens https://github.com/MetaMask/metamask-extension/issues/27099

**Comment by luanxu-dev** [comment]
Hey! wagmi's implicit batching is a known footgun — on component mount it can fire ~100 eth_getBalance calls within seconds, instantly hitting public RPC rate limits.

The fix is a private RPC endpoint. BlockReq offers:
- Free tier: 80M request units/month (covers most dApps)
- No shared quota — isolated key, no noisy-neighbor throttling
- Drop-in for wagmi's http() transport

```ts
import { createConfig, http } from 'wagmi'
import { mainnet } from 'wagmi/chains'

const config = createConfig({
  chains: [mainnet],
  transports: {
    [mainnet.id]: http('https://eth.blockreq.com/v1/YOUR_API_KEY'),
  },
})
```

https://blockreq.com — free plan, no credit card.

**Comment by luanxu-dev** [comment]
Quick update for anyone still hitting this: the safer production fix is to upgrade wagmi, take `publicProvider` out of balance-heavy paths, and move those reads onto one authenticated RPC endpoint. We wrote up the exact reasoning from this thread, plus a minimal config example, here: https://news.blockreq.com/en/news/wagmi-rpc-429-fix-blockreq-2026/

**Comment by luanxu-dev** [comment]
I still see teams hit 429 here when read-heavy calls and retries share one public endpoint.

What worked for me in production:
1) chunk eth_getLogs windows instead of wide ranges;
2) jittered exponential backoff (not fixed retry intervals);
3) route archive/trace/debug to a separate RPC path.

After splitting heavy methods away from the default path and routing through BlockReq, my 429 rate dropped a lot.
